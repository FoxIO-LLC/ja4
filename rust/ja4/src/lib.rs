// Copyright (c) 2023, FoxIO, LLC.
// All rights reserved.
// Patent Pending
// JA4 is Open-Source, Licensed under BSD 3-Clause
// JA4+ (JA4S, JA4H, JA4L, JA4X, JA4SSH) are licenced under the FoxIO License 1.1.
// For full license text, see the repo root.

mod conf;
mod error;
mod http;
mod pcap;
mod ssh;
mod stream;
mod time;
mod tls;

use std::{io::Write, path::PathBuf};

use clap::Parser;
use rtshark::RTSharkBuilder;

pub use crate::error::Error;
use crate::{
    conf::Conf,
    pcap::{Packet, PacketNum, Proto},
    stream::Streams,
};

pub type Result<T, E = Error> = std::result::Result<T, E>;

/// Calculate JA4 fingerprints
#[derive(Debug, Parser)]
#[command(version = env!("CARGO_PKG_VERSION"))]
pub struct Cli {
    /// JSON output (default is YAML)
    #[arg(short, long)]
    json: bool,
    /// Include raw (unhashed) fingerprints in the output
    #[arg(short = 'r', long)]
    with_raw: bool,
    /// Preserve the original order of values.
    ///
    /// JA4 (TLS client): disable sorting of ciphers and TLS extensions.
    ///
    /// JA4H (HTTP client): disable sorting of headers and cookies.
    #[arg(short = 'O', long)]
    original_order: bool,
    /// The key log file that enables decryption of TLS traffic.
    ///
    /// This file is generated by the browser when `SSLKEYLOGFILE` environment variable is set.
    /// See <https://wiki.wireshark.org/TLS#using-the-pre-master-secret> for more details.
    ///
    /// Note that you can embed the TLS key log file in a capture file:
    /// `editcap --inject-secrets tls,keys.txt in.pcap out-dsb.pcapng`
    #[arg(long)]
    keylog_file: Option<PathBuf>,
    /// Include packet numbers (`pkt_*` fields) in the output.
    ///
    /// This information is useful for debugging.
    #[arg(short = 'n', long)]
    with_packet_numbers: bool,
    /// The capture file to process
    pcap: PathBuf,
}

impl Cli {
    /// Write JSON with JA4 fingerprints to the I/O stream.
    pub fn run<W: Write>(self, writer: &mut W) -> Result<()> {
        let conf = Conf::load()?;
        let Cli {
            json,
            with_raw,
            original_order,
            keylog_file,
            with_packet_numbers,
            pcap,
        } = self;

        let Some(pcap_path) = pcap.to_str() else {
            return Err(Error::NonUtf8Path(pcap));
        };
        check_tshark_version()?;
        let mut builder = RTSharkBuilder::builder().input_path(pcap_path);

        if let Some(keylog) = &keylog_file {
            let Some(keylog_path) = keylog.to_str() else {
                // SAFETY: we've just established that `keylog_file` is Some
                return Err(Error::NonUtf8Path(keylog_file.unwrap()));
            };
            builder = builder.keylog_file(keylog_path);
        }
        let mut tshark = builder.spawn()?;

        let mut streams = Streams::default();

        let mut packet_num = 0;
        while let Some(packet) = tshark.read().unwrap_or_else(|err| {
            tracing::error!(%err, "failed to parse tshark output");
            None
        }) {
            packet_num += 1;
            let pkt = Packet::new(&packet, packet_num);

            if let Err(error) = streams.update(&pkt, &conf, with_packet_numbers) {
                tracing::debug!(packet_num, %error, "failed to handle packet");
            }
        }

        let flags = FormatFlags {
            with_raw,
            original_order,
        };
        // HACK: The purpose of the `io::stdout` mumbo-jumbo is to handle
        // BrokenPipe error. Rust throws it when the stdout is piped to `head`.
        if json {
            for rec in streams.into_out(flags) {
                serde_json::to_writer(&mut *writer, &rec)?;
                writeln!(writer)?;
            }
        } else {
            let s = serde_yaml::to_string(&streams.into_out(flags).collect::<Vec<_>>())?;
            writer.write_all(s.as_bytes())?;
        }
        Ok(())
    }
}

#[derive(Debug, Clone, Copy, Default)]
pub(crate) struct FormatFlags {
    /// Whether to add raw (unhashed) fingerprints to the output.
    ///
    /// Corresponds to [`Cli::with_raw`].
    pub(crate) with_raw: bool,
    /// Whether to preserve the original order of values.
    ///
    /// Corresponds to [`Cli::original_order`].
    pub(crate) original_order: bool,
}

/// Which side of the connection sent the packet?
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum Sender {
    Client,
    Server,
}

/// Returns first 12 characters of the SHA-256 hash of the given string.
///
/// Returns `"000000000000"` (12 zeros) if the input string is empty.
fn hash12(s: impl AsRef<str>) -> String {
    use sha2::{Digest as _, Sha256};

    let s = s.as_ref();
    if s.is_empty() {
        "000000000000".to_owned()
    } else {
        let sha256 = hex::encode(Sha256::digest(s));
        sha256[..12].into()
    }
}

#[test]
fn test_hash12() {
    assert_eq!(hash12("551d0f,551d25,551d11"), "aae71e8db6d7");
    assert_eq!(hash12(""), "000000000000");
}

fn check_tshark_version() -> Result<()> {
    use owo_colors::OwoColorize as _;

    let out = duct::cmd!("tshark", "--version")
        .read()
        .map_err(|e| Error::TsharkNotFound { source: e })?;
    tracing::debug!(%out, "tshark --version");

    let ver = parse_tshark_version(&out).ok_or(Error::ParseTsharkVersion)?;
    let available = semver::Version::parse(ver)?;

    let required = semver::VersionReq::parse(">=4.0.6").expect("BUG");
    if !required.matches(&available) {
        tracing::warn!(%available, %required, "tshark version is outdated");
        let warning = format!(
            "⚠️  You are running an older version of tshark ({available}).\n\
            JA4 is designed to work with tshark version 4.0.6 and above.\n\
            Some functionality may not work properly with older versions."
        );
        eprintln!("{}", warning.bold().red());
    }
    Ok(())
}

/// Parses the version number from the output of `tshark --version`.
fn parse_tshark_version(tshark_version_output: &str) -> Option<&str> {
    // The first line of `tshark --version` output is formatted like this:
    // "TShark (Wireshark) 4.0.8 (v4.0.8-0-g81696bb74857).\n"
    let start = tshark_version_output.find(") ").map(|i| i + 2)?;
    let version_start = &tshark_version_output[start..];
    let end = version_start.find(char::is_whitespace)?;
    let ver = &version_start[..end];
    Some(ver.strip_suffix('.').unwrap_or(ver))
}

#[test]
fn test_parse_tshark_version() {
    assert_eq!(
        parse_tshark_version("TShark (Wireshark) 4.0.8 (v4.0.8-0-g81696bb74857)."),
        Some("4.0.8")
    );
    assert_eq!(
        parse_tshark_version("TShark (Wireshark) 3.6.2 (Git v3.6.2 packaged as 3.6.2-2)"),
        Some("3.6.2")
    );
    assert_eq!(
        parse_tshark_version("TShark (Wireshark) 4.4.0.\n\nCopyright 1998-2024"),
        Some("4.4.0")
    );
    // Abrupt end of the string.
    assert!(parse_tshark_version("TShark (Wireshark) 4.4.0.").is_none());
    assert!(parse_tshark_version("What the TShark?!").is_none());
}

// XXX-FIXME(vvv): `test_insta` fails on Windows; see https://github.com/FoxIO-LLC/ja4/issues/10
#[cfg(not(windows))]
#[test]
fn test_insta() {
    insta::glob!(
        concat!(env!("CARGO_MANIFEST_DIR"), "/../.."),
        "pcap/*.pcap*",
        |path| {
            let cli = Cli {
                json: false,
                with_raw: false,
                original_order: false,
                keylog_file: None,
                with_packet_numbers: false,
                pcap: path.to_path_buf(),
            };

            let mut output = Vec::<u8>::new();
            cli.run(&mut output).unwrap();
            let output = String::from_utf8(output).unwrap();

            insta::assert_snapshot!(output);
        }
    );
}
