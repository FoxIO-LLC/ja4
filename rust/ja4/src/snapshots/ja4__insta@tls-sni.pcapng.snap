---
source: ja4/src/lib.rs
expression: output
---
- stream: 0
  transport: tcp
  src: 192.168.1.168
  dst: 142.251.16.94
  src_port: 50112
  dst_port: 443
  tls_server_name: clientservices.googleapis.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 1
  transport: tcp
  src: 192.168.1.168
  dst: 142.251.163.147
  src_port: 50113
  dst_port: 443
  tls_server_name: www.google.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 2
  transport: tcp
  src: 192.168.1.168
  dst: 172.253.122.84
  src_port: 50114
  dst_port: 443
  tls_server_name: accounts.google.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 3
  transport: tcp
  src: 192.168.1.168
  dst: 142.251.16.95
  src_port: 50115
  dst_port: 443
  tls_server_name: www.googleapis.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 4
  transport: tcp
  src: 192.168.1.168
  dst: 104.112.30.74
  src_port: 50116
  dst_port: 443
  tls_server_name: lastpass.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 5
  transport: tcp
  src: 192.168.1.168
  dst: 192.241.241.147
  src_port: 50122
  dst_port: 443
  tls_server_name: lp-push-server-452.lastpass.com
  ja4: t13d1516h1_8daaf6152771_e5627efa2ab1
- stream: 6
  transport: tcp
  src: 192.168.1.168
  dst: 142.251.163.188
  src_port: 50123
  dst_port: 5228
  tls_server_name: mtalk.google.com
  ja4: t13d151400_8daaf6152771_de4a06bb82e3
- stream: 7
  transport: tcp
  src: 192.168.1.168
  dst: 172.253.122.94
  src_port: 50126
  dst_port: 443
  tls_server_name: www.gstatic.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 8
  transport: tcp
  src: 192.168.1.168
  dst: 142.251.16.100
  src_port: 50127
  dst_port: 443
  tls_server_name: ogs.google.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 9
  transport: tcp
  src: 192.168.1.168
  dst: 142.251.111.101
  src_port: 50128
  dst_port: 443
  tls_server_name: aa.google.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 10
  transport: tcp
  src: 192.168.1.168
  dst: 142.251.163.95
  src_port: 50130
  dst_port: 443
  tls_server_name: safebrowsing.googleapis.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 11
  transport: tcp
  src: 192.168.1.168
  dst: 31.13.66.35
  src_port: 50131
  dst_port: 443
  tls_server_name: www.facebook.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 12
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.1
  src_port: 50134
  dst_port: 443
  tls_server_name: static.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 13
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.1
  src_port: 50132
  dst_port: 443
  tls_server_name: static.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 14
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.1
  src_port: 50135
  dst_port: 443
  tls_server_name: static.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 15
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.1
  src_port: 50133
  dst_port: 443
  tls_server_name: static.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 16
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.1
  src_port: 50136
  dst_port: 443
  tls_server_name: static.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 17
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.1
  src_port: 50137
  dst_port: 443
  tls_server_name: static.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 18
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.1
  src_port: 50139
  dst_port: 443
  tls_server_name: scontent-iad3-2.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 19
  transport: tcp
  src: 192.168.1.168
  dst: 31.13.66.2
  src_port: 50140
  dst_port: 443
  tls_server_name: gateway.facebook.com
  ja4: t13d1516h1_8daaf6152771_e5627efa2ab1
- stream: 20
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.241.1
  src_port: 50142
  dst_port: 443
  tls_server_name: scontent-lga3-2.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 21
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.14.19
  src_port: 50141
  dst_port: 443
  tls_server_name: scontent-mia3-2.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 22
  transport: tcp
  src: 192.168.1.168
  dst: 208.255.115.145
  src_port: 50143
  dst_port: 443
  tls_server_name: scontent.fewr1-6.fna.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 23
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.2
  src_port: 50145
  dst_port: 443
  tls_server_name: video-iad3-2.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 24
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.2
  src_port: 50146
  dst_port: 443
  tls_server_name: video-iad3-2.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 25
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.2
  src_port: 50144
  dst_port: 443
  tls_server_name: video-iad3-2.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 26
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.2
  src_port: 50147
  dst_port: 443
  tls_server_name: video-iad3-2.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 27
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.2
  src_port: 50148
  dst_port: 443
  tls_server_name: video-iad3-2.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 28
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.2
  src_port: 50149
  dst_port: 443
  tls_server_name: video-iad3-2.xx.fbcdn.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 29
  transport: tcp
  src: 192.168.1.168
  dst: 142.251.163.95
  src_port: 50151
  dst_port: 443
  tls_server_name: content-autofill.googleapis.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 30
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.17
  src_port: 50152
  dst_port: 443
  tls_server_name: edge-chat.facebook.com
  ja4: t13d1516h1_8daaf6152771_e5627efa2ab1
- stream: 31
  transport: tcp
  src: 192.168.1.168
  dst: 157.240.229.17
  src_port: 50153
  dst_port: 443
  tls_server_name: edge-chat.facebook.com
  ja4: t13d1517h1_8daaf6152771_6cdcb247c39b
- stream: 32
  transport: tcp
  src: 192.168.1.168
  dst: 23.50.125.163
  src_port: 50155
  dst_port: 443
  tls_server_name: www.microsoft.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 33
  transport: tcp
  src: 192.168.1.168
  dst: 23.218.218.147
  src_port: 50157
  dst_port: 443
  tls_server_name: statics-marketingsites-eus-ms-com.akamaized.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 34
  transport: tcp
  src: 192.168.1.168
  dst: 23.218.218.171
  src_port: 50158
  dst_port: 443
  tls_server_name: img-prod-cms-rt-microsoft-com.akamaized.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 35
  transport: tcp
  src: 192.168.1.168
  dst: 23.212.251.12
  src_port: 50160
  dst_port: 443
  tls_server_name: cdn-dynmedia-1.microsoft.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 36
  transport: tcp
  src: 192.168.1.168
  dst: 23.212.251.12
  src_port: 50162
  dst_port: 443
  tls_server_name: cdn-dynmedia-1.microsoft.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 37
  transport: tcp
  src: 192.168.1.168
  dst: 23.212.251.12
  src_port: 50161
  dst_port: 443
  tls_server_name: cdn-dynmedia-1.microsoft.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 38
  transport: tcp
  src: 192.168.1.168
  dst: 13.107.237.40
  src_port: 50159
  dst_port: 443
  tls_server_name: mem.gfx.ms
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 39
  transport: tcp
  src: 192.168.1.168
  dst: 18.67.65.105
  src_port: 50163
  dst_port: 443
  tls_server_name: via.placeholder.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 40
  transport: tcp
  src: 192.168.1.168
  dst: 13.107.237.40
  src_port: 50164
  dst_port: 443
  tls_server_name: wcpstatic.microsoft.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 41
  transport: tcp
  src: 192.168.1.168
  dst: 13.107.238.40
  src_port: 50165
  dst_port: 443
  tls_server_name: js.monitor.azure.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 42
  transport: tcp
  src: 192.168.1.168
  dst: 13.107.237.40
  src_port: 50166
  dst_port: 443
  tls_server_name: mem.gfx.ms
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 43
  transport: tcp
  src: 192.168.1.168
  dst: 40.126.24.84
  src_port: 50167
  dst_port: 443
  tls_server_name: login.live.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 44
  transport: tcp
  src: 192.168.1.168
  dst: 13.107.238.40
  src_port: 50168
  dst_port: 443
  tls_server_name: logincdn.msauth.net
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 45
  transport: tcp
  src: 192.168.1.168
  dst: 13.107.237.40
  src_port: 50169
  dst_port: 443
  tls_server_name: mem.gfx.ms
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 46
  transport: tcp
  src: 192.168.1.168
  dst: 3.223.179.120
  src_port: 50170
  dst_port: 443
  tls_server_name: target.microsoft.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 47
  transport: tcp
  src: 192.168.1.168
  dst: 23.55.200.211
  src_port: 50172
  dst_port: 443
  tls_server_name: www.apple.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 48
  transport: tcp
  src: 192.168.1.168
  dst: 23.62.168.26
  src_port: 50175
  dst_port: 443
  tls_server_name: is1-ssl.mzstatic.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 49
  transport: tcp
  src: 192.168.1.168
  dst: 23.62.168.26
  src_port: 50174
  dst_port: 443
  tls_server_name: is1-ssl.mzstatic.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 50
  transport: tcp
  src: 192.168.1.168
  dst: 23.62.168.26
  src_port: 50177
  dst_port: 443
  tls_server_name: is2-ssl.mzstatic.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 51
  transport: tcp
  src: 192.168.1.168
  dst: 23.62.168.26
  src_port: 50176
  dst_port: 443
  tls_server_name: is2-ssl.mzstatic.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 52
  transport: tcp
  src: 192.168.1.168
  dst: 23.62.168.26
  src_port: 50178
  dst_port: 443
  tls_server_name: is2-ssl.mzstatic.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 53
  transport: tcp
  src: 192.168.1.168
  dst: 23.62.168.26
  src_port: 50180
  dst_port: 443
  tls_server_name: is3-ssl.mzstatic.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 54
  transport: tcp
  src: 192.168.1.168
  dst: 23.62.168.26
  src_port: 50182
  dst_port: 443
  tls_server_name: is3-ssl.mzstatic.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 55
  transport: tcp
  src: 192.168.1.168
  dst: 23.62.168.26
  src_port: 50181
  dst_port: 443
  tls_server_name: is3-ssl.mzstatic.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 56
  transport: tcp
  src: 192.168.1.168
  dst: 23.62.168.26
  src_port: 50179
  dst_port: 443
  tls_server_name: is2-ssl.mzstatic.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 57
  transport: tcp
  src: 192.168.1.168
  dst: 23.62.168.26
  src_port: 50183
  dst_port: 443
  tls_server_name: is5-ssl.mzstatic.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 58
  transport: tcp
  src: 192.168.1.168
  dst: 172.253.63.95
  src_port: 50184
  dst_port: 443
  tls_server_name: content-autofill.googleapis.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
- stream: 0
  transport: udp
  src: 192.168.1.168
  dst: 142.251.163.147
  src_port: 59102
  dst_port: 443
  tls_server_name: www.google.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 1
  transport: udp
  src: 192.168.1.168
  dst: 172.253.122.84
  src_port: 57816
  dst_port: 443
  tls_server_name: accounts.google.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 2
  transport: udp
  src: 192.168.1.168
  dst: 142.251.16.95
  src_port: 55945
  dst_port: 443
  tls_server_name: www.googleapis.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 3
  transport: udp
  src: 192.168.1.168
  dst: 142.251.163.94
  src_port: 49333
  dst_port: 443
  tls_server_name: update.googleapis.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 4
  transport: udp
  src: 192.168.1.168
  dst: 142.251.163.102
  src_port: 57662
  dst_port: 443
  tls_server_name: encrypted-tbn0.gstatic.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 5
  transport: udp
  src: 192.168.1.168
  dst: 142.251.111.101
  src_port: 60486
  dst_port: 443
  tls_server_name: history.google.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 6
  transport: udp
  src: 192.168.1.168
  dst: 172.253.122.94
  src_port: 60845
  dst_port: 443
  tls_server_name: www.gstatic.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 7
  transport: udp
  src: 192.168.1.168
  dst: 142.251.16.100
  src_port: 55906
  dst_port: 443
  tls_server_name: ogs.google.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 8
  transport: udp
  src: 192.168.1.168
  dst: 142.251.111.101
  src_port: 61836
  dst_port: 443
  tls_server_name: aa.google.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 9
  transport: udp
  src: 192.168.1.168
  dst: 172.253.63.95
  src_port: 59221
  dst_port: 443
  tls_server_name: content-autofill.googleapis.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 10
  transport: udp
  src: 192.168.1.168
  dst: 142.251.163.138
  src_port: 51905
  dst_port: 443
  tls_server_name: encrypted-tbn0.gstatic.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 11
  transport: udp
  src: 192.168.1.168
  dst: 142.251.163.95
  src_port: 51782
  dst_port: 443
  tls_server_name: safebrowsing.googleapis.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 12
  transport: udp
  src: 192.168.1.168
  dst: 157.240.229.1
  src_port: 61730
  dst_port: 443
  tls_server_name: scontent-iad3-2.xx.fbcdn.net
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 13
  transport: udp
  src: 192.168.1.168
  dst: 157.240.229.1
  src_port: 54980
  dst_port: 443
  tls_server_name: static.xx.fbcdn.net
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 14
  transport: udp
  src: 192.168.1.168
  dst: 31.13.66.35
  src_port: 52166
  dst_port: 443
  tls_server_name: www.facebook.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 15
  transport: udp
  src: 192.168.1.168
  dst: 157.240.229.1
  src_port: 59444
  dst_port: 443
  tls_server_name: static.xx.fbcdn.net
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 16
  transport: udp
  src: 192.168.1.168
  dst: 157.240.229.2
  src_port: 60374
  dst_port: 443
  tls_server_name: video-iad3-2.xx.fbcdn.net
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 17
  transport: udp
  src: 192.168.1.168
  dst: 142.251.163.95
  src_port: 55309
  dst_port: 443
  tls_server_name: content-autofill.googleapis.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 18
  transport: udp
  src: 192.168.1.168
  dst: 142.251.163.95
  src_port: 50053
  dst_port: 443
  tls_server_name: optimizationguide-pa.googleapis.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918
- stream: 19
  transport: udp
  src: 192.168.1.168
  dst: 142.251.163.132
  src_port: 57188
  dst_port: 443
  tls_server_name: lh5.googleusercontent.com
  ja4: q13d0310h3_55b375c5d22e_cd85d2d88918

